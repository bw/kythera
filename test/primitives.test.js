const t = require("./util").test

describe("Primitive Types", () => {
	describe("Initialization with 'new'", () => {
		t("bool", `new bool`)
		t("int", `new int`)
		t("float", `new float`)
		t("str", `new str`)
		t("null", `new null`)
	})

	describe("Declaration and assignment", () => {
		t("bool", `
let myBool = false
myBool = true
		`)

		t("int", `
let myInt = 99
myInt = 3
`)
		t("float", `
let myFloat = 3.14159
myFloat = 2.71828183
`)
		t("str", `
let myStr = "hello"
myStr = "world"
		`)

		t("null", `
let myNull = null
myNull = null
	`)
	})

	describe("Literals", () => {
		t("true", `true`)
		t("false", `false`)
		t("int", `10`)
		t("float", `3.14159`)
		t("str", `"kythera"`)
		t("null", `null`)
	})

	describe("Arithmetic", () => {
		t("int Addition", `2+2`)
		t("float addition", `3.14159 + 2.71828`)
		t("subtraction", `9-5`)
		t("division", `6 / 3`)
		t("multiplication", `4 * 2`)
		t("modulo", `18 % 3`)
	})

	describe("Boolean operations", () => {
		t("negate false", `!false`, {compile: false})
		t("negate true", `!true`, {compile: false})
		t("and", `true && true`)
		t("and", `true && false`)
		t("short-circuit and", `false && true`)
		t("or", `false || false`)
		t("or", `false || true`)
		t("short-circuit or", `true || false`)
	})

	describe("typeof", () => {
		t("true", `typeof true`)
		t("false", `typeof false`)
		t("int", `typeof 10`)
		t("float", `typeof 3.14159`)
		t("str", `typeof "kythera"`)
		t("null", `typeof null`)
	})

	describe("Casting", () => {
		describe("ints and floats", () => {
			t("int as float", `10 as float`, {compile: false})
			t("float as int", `3.14159 as int`, {compile: false})
		})

		describe("ints and floats to bool", () => {
			t("1.0 as bool", `1.0 as bool`, {compile: false})
			t("0.0 as bool", `0.0 as bool`, {compile: false})
			t("1 as bool", `1 as bool`, {compile: false})
			t("0 as bool", `0 as bool`, {compile: false})
		})
	})
})