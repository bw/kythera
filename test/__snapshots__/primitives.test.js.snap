// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Primitive Types Arithmetic [COMPL] division 1`] = `"(new KYTHERA.value(6, KYTHERA.type.PRIMITIVES[\\"int\\"])) / (new KYTHERA.value(3, KYTHERA.type.PRIMITIVES[\\"int\\"]));"`;

exports[`Primitive Types Arithmetic [COMPL] float addition 1`] = `"(new KYTHERA.value(3.14159, KYTHERA.type.PRIMITIVES[\\"float\\"])) + (new KYTHERA.value(2.71828, KYTHERA.type.PRIMITIVES[\\"float\\"]));"`;

exports[`Primitive Types Arithmetic [COMPL] int Addition 1`] = `"(new KYTHERA.value(2, KYTHERA.type.PRIMITIVES[\\"int\\"])) + (new KYTHERA.value(2, KYTHERA.type.PRIMITIVES[\\"int\\"]));"`;

exports[`Primitive Types Arithmetic [COMPL] modulo 1`] = `"(new KYTHERA.value(18, KYTHERA.type.PRIMITIVES[\\"int\\"])) % (new KYTHERA.value(3, KYTHERA.type.PRIMITIVES[\\"int\\"]));"`;

exports[`Primitive Types Arithmetic [COMPL] multiplication 1`] = `"(new KYTHERA.value(4, KYTHERA.type.PRIMITIVES[\\"int\\"])) * (new KYTHERA.value(2, KYTHERA.type.PRIMITIVES[\\"int\\"]));"`;

exports[`Primitive Types Arithmetic [COMPL] subtraction 1`] = `"(new KYTHERA.value(9, KYTHERA.type.PRIMITIVES[\\"int\\"])) - (new KYTHERA.value(5, KYTHERA.type.PRIMITIVES[\\"int\\"]));"`;

exports[`Primitive Types Arithmetic [PARSE] division 1`] = `
Array [
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 6,
    },
    "operator": "/",
    "right": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 3,
    },
  },
]
`;

exports[`Primitive Types Arithmetic [PARSE] float addition 1`] = `
Array [
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "float",
      },
      "value": 3.14159,
    },
    "operator": "+",
    "right": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "float",
      },
      "value": 2.71828,
    },
  },
]
`;

exports[`Primitive Types Arithmetic [PARSE] int Addition 1`] = `
Array [
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 2,
    },
    "operator": "+",
    "right": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 2,
    },
  },
]
`;

exports[`Primitive Types Arithmetic [PARSE] modulo 1`] = `
Array [
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 18,
    },
    "operator": "%",
    "right": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 3,
    },
  },
]
`;

exports[`Primitive Types Arithmetic [PARSE] multiplication 1`] = `
Array [
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 4,
    },
    "operator": "*",
    "right": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 2,
    },
  },
]
`;

exports[`Primitive Types Arithmetic [PARSE] subtraction 1`] = `
Array [
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 9,
    },
    "operator": "-",
    "right": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 5,
    },
  },
]
`;

exports[`Primitive Types Boolean operations [COMPL] and 1`] = `"(new KYTHERA.value(true, KYTHERA.type.PRIMITIVES[\\"bool\\"])) && (new KYTHERA.value(true, KYTHERA.type.PRIMITIVES[\\"bool\\"]));"`;

exports[`Primitive Types Boolean operations [COMPL] and 2`] = `"(new KYTHERA.value(true, KYTHERA.type.PRIMITIVES[\\"bool\\"])) && (new KYTHERA.value(false, KYTHERA.type.PRIMITIVES[\\"bool\\"]));"`;

exports[`Primitive Types Boolean operations [COMPL] or 1`] = `"(new KYTHERA.value(false, KYTHERA.type.PRIMITIVES[\\"bool\\"])) || (new KYTHERA.value(false, KYTHERA.type.PRIMITIVES[\\"bool\\"]));"`;

exports[`Primitive Types Boolean operations [COMPL] or 2`] = `"(new KYTHERA.value(false, KYTHERA.type.PRIMITIVES[\\"bool\\"])) || (new KYTHERA.value(true, KYTHERA.type.PRIMITIVES[\\"bool\\"]));"`;

exports[`Primitive Types Boolean operations [COMPL] short-circuit and 1`] = `"(new KYTHERA.value(false, KYTHERA.type.PRIMITIVES[\\"bool\\"])) && (new KYTHERA.value(true, KYTHERA.type.PRIMITIVES[\\"bool\\"]));"`;

exports[`Primitive Types Boolean operations [COMPL] short-circuit or 1`] = `"(new KYTHERA.value(true, KYTHERA.type.PRIMITIVES[\\"bool\\"])) || (new KYTHERA.value(false, KYTHERA.type.PRIMITIVES[\\"bool\\"]));"`;

exports[`Primitive Types Boolean operations [PARSE] and 1`] = `
Array [
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": true,
    },
    "operator": "&&",
    "right": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": true,
    },
  },
]
`;

exports[`Primitive Types Boolean operations [PARSE] and 2`] = `
Array [
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": true,
    },
    "operator": "&&",
    "right": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": false,
    },
  },
]
`;

exports[`Primitive Types Boolean operations [PARSE] negate false 1`] = `
Array [
  ParseNode {
    "kind": "unary",
    "operator": "!",
    "target": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": false,
    },
  },
]
`;

exports[`Primitive Types Boolean operations [PARSE] negate true 1`] = `
Array [
  ParseNode {
    "kind": "unary",
    "operator": "!",
    "target": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": true,
    },
  },
]
`;

exports[`Primitive Types Boolean operations [PARSE] or 1`] = `
Array [
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": false,
    },
    "operator": "||",
    "right": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": false,
    },
  },
]
`;

exports[`Primitive Types Boolean operations [PARSE] or 2`] = `
Array [
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": false,
    },
    "operator": "||",
    "right": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": true,
    },
  },
]
`;

exports[`Primitive Types Boolean operations [PARSE] short-circuit and 1`] = `
Array [
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": false,
    },
    "operator": "&&",
    "right": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": true,
    },
  },
]
`;

exports[`Primitive Types Boolean operations [PARSE] short-circuit or 1`] = `
Array [
  ParseNode {
    "kind": "binary",
    "left": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": true,
    },
    "operator": "||",
    "right": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": false,
    },
  },
]
`;

exports[`Primitive Types Casting ints and floats [PARSE] float as int 1`] = `
Array [
  ParseNode {
    "from": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "float",
      },
      "value": 3.14159,
    },
    "kind": "as",
    "to": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "int",
    },
  },
]
`;

exports[`Primitive Types Casting ints and floats [PARSE] int as float 1`] = `
Array [
  ParseNode {
    "from": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 10,
    },
    "kind": "as",
    "to": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "float",
    },
  },
]
`;

exports[`Primitive Types Casting ints and floats to bool [PARSE] 0 as bool 1`] = `
Array [
  ParseNode {
    "from": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 0,
    },
    "kind": "as",
    "to": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "bool",
    },
  },
]
`;

exports[`Primitive Types Casting ints and floats to bool [PARSE] 0.0 as bool 1`] = `
Array [
  ParseNode {
    "from": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 0,
    },
    "kind": "as",
    "to": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "bool",
    },
  },
]
`;

exports[`Primitive Types Casting ints and floats to bool [PARSE] 1 as bool 1`] = `
Array [
  ParseNode {
    "from": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 1,
    },
    "kind": "as",
    "to": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "bool",
    },
  },
]
`;

exports[`Primitive Types Casting ints and floats to bool [PARSE] 1.0 as bool 1`] = `
Array [
  ParseNode {
    "from": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 1,
    },
    "kind": "as",
    "to": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "bool",
    },
  },
]
`;

exports[`Primitive Types Initialization with 'new' [COMPL] bool 1`] = `"KYTHERA.type.PRIMITIVES[\\"bool\\"].makeNew();"`;

exports[`Primitive Types Initialization with 'new' [COMPL] float 1`] = `"KYTHERA.type.PRIMITIVES[\\"float\\"].makeNew();"`;

exports[`Primitive Types Initialization with 'new' [COMPL] int 1`] = `"KYTHERA.type.PRIMITIVES[\\"int\\"].makeNew();"`;

exports[`Primitive Types Initialization with 'new' [COMPL] null 1`] = `"KYTHERA.type.PRIMITIVES[\\"null\\"].makeNew();"`;

exports[`Primitive Types Initialization with 'new' [COMPL] str 1`] = `"KYTHERA.type.PRIMITIVES[\\"str\\"].makeNew();"`;

exports[`Primitive Types Initialization with 'new' [PARSE] bool 1`] = `
Array [
  ParseNode {
    "kind": "new",
    "target": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "bool",
    },
  },
]
`;

exports[`Primitive Types Initialization with 'new' [PARSE] float 1`] = `
Array [
  ParseNode {
    "kind": "new",
    "target": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "float",
    },
  },
]
`;

exports[`Primitive Types Initialization with 'new' [PARSE] int 1`] = `
Array [
  ParseNode {
    "kind": "new",
    "target": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "int",
    },
  },
]
`;

exports[`Primitive Types Initialization with 'new' [PARSE] null 1`] = `
Array [
  ParseNode {
    "kind": "new",
    "target": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "null",
    },
  },
]
`;

exports[`Primitive Types Initialization with 'new' [PARSE] str 1`] = `
Array [
  ParseNode {
    "kind": "new",
    "target": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "str",
    },
  },
]
`;

exports[`Primitive Types Literals [COMPL] false 1`] = `"new KYTHERA.value(false, KYTHERA.type.PRIMITIVES[\\"bool\\"]);"`;

exports[`Primitive Types Literals [COMPL] float 1`] = `"new KYTHERA.value(3.14159, KYTHERA.type.PRIMITIVES[\\"float\\"]);"`;

exports[`Primitive Types Literals [COMPL] int 1`] = `"new KYTHERA.value(10, KYTHERA.type.PRIMITIVES[\\"int\\"]);"`;

exports[`Primitive Types Literals [COMPL] null 1`] = `"new KYTHERA.value(null, KYTHERA.type.PRIMITIVES[\\"null\\"]);"`;

exports[`Primitive Types Literals [COMPL] str 1`] = `"new KYTHERA.value(\\"kythera\\", KYTHERA.type.PRIMITIVES[\\"str\\"]);"`;

exports[`Primitive Types Literals [COMPL] true 1`] = `"new KYTHERA.value(true, KYTHERA.type.PRIMITIVES[\\"bool\\"]);"`;

exports[`Primitive Types Literals [PARSE] false 1`] = `
Array [
  ParseNode {
    "kind": "literal",
    "type": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "bool",
    },
    "value": false,
  },
]
`;

exports[`Primitive Types Literals [PARSE] float 1`] = `
Array [
  ParseNode {
    "kind": "literal",
    "type": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "float",
    },
    "value": 3.14159,
  },
]
`;

exports[`Primitive Types Literals [PARSE] int 1`] = `
Array [
  ParseNode {
    "kind": "literal",
    "type": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "int",
    },
    "value": 10,
  },
]
`;

exports[`Primitive Types Literals [PARSE] null 1`] = `
Array [
  ParseNode {
    "kind": "literal",
    "type": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "null",
    },
    "value": null,
  },
]
`;

exports[`Primitive Types Literals [PARSE] str 1`] = `
Array [
  ParseNode {
    "kind": "literal",
    "type": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "str",
    },
    "value": "kythera",
  },
]
`;

exports[`Primitive Types Literals [PARSE] true 1`] = `
Array [
  ParseNode {
    "kind": "literal",
    "type": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "bool",
    },
    "value": true,
  },
]
`;

exports[`Primitive Types typeof [COMPL] false 1`] = `"KYTHERA.type.PRIMITIVES[\\"bool\\"];"`;

exports[`Primitive Types typeof [COMPL] float 1`] = `"KYTHERA.type.PRIMITIVES[\\"float\\"];"`;

exports[`Primitive Types typeof [COMPL] int 1`] = `"KYTHERA.type.PRIMITIVES[\\"int\\"];"`;

exports[`Primitive Types typeof [COMPL] null 1`] = `"KYTHERA.type.PRIMITIVES[\\"null\\"];"`;

exports[`Primitive Types typeof [COMPL] str 1`] = `"KYTHERA.type.PRIMITIVES[\\"str\\"];"`;

exports[`Primitive Types typeof [COMPL] true 1`] = `"KYTHERA.type.PRIMITIVES[\\"bool\\"];"`;

exports[`Primitive Types typeof [PARSE] false 1`] = `
Array [
  ParseNode {
    "kind": "typeof",
    "target": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": false,
    },
  },
]
`;

exports[`Primitive Types typeof [PARSE] float 1`] = `
Array [
  ParseNode {
    "kind": "typeof",
    "target": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "float",
      },
      "value": 3.14159,
    },
  },
]
`;

exports[`Primitive Types typeof [PARSE] int 1`] = `
Array [
  ParseNode {
    "kind": "typeof",
    "target": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 10,
    },
  },
]
`;

exports[`Primitive Types typeof [PARSE] null 1`] = `
Array [
  ParseNode {
    "kind": "typeof",
    "target": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "null",
      },
      "value": null,
    },
  },
]
`;

exports[`Primitive Types typeof [PARSE] str 1`] = `
Array [
  ParseNode {
    "kind": "typeof",
    "target": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "str",
      },
      "value": "kythera",
    },
  },
]
`;

exports[`Primitive Types typeof [PARSE] true 1`] = `
Array [
  ParseNode {
    "kind": "typeof",
    "target": ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": true,
    },
  },
]
`;
